pageserver/src/tenant/timeline/init.rs

```diff
@@ -104,26 +118,38 @@ pub(super) enum DismissedLayer {
 
 /// Merges local discoveries and remote [`IndexPart`] to a collection of decisions.
 pub(super) fn reconcile(
-    discovered: Vec<(LayerFileName, u64)>,
+    discovered: Vec<(LayerFileName, Utf8PathBuf, u64)>,
     index_part: Option<&IndexPart>,
     disk_consistent_lsn: Lsn,
     generation: Generation,
     shard: ShardIndex,
-) -> Vec<(LayerFileName, Result<Decision, DismissedLayer>)> {
+) -> Vec<(
+    LayerFileName,
+    Option<Utf8PathBuf>,
+    Result<Decision, DismissedLayer>,
+)> {
     use Decision::*;
 
-    // name => (local, remote)
-    type Collected = HashMap<LayerFileName, (Option<LayerFileMetadata>, Option<LayerFileMetadata>)>;
+    // name => (local_path, local_metadata, remote_metadata)
+    type Collected = HashMap<
+        LayerFileName,
+        (
+            Option<Utf8PathBuf>,
+            Option<LayerFileMetadata>,
+            Option<LayerFileMetadata>,
+        ),
+    >;
```

Comment:
Both the 3-tuple in the returned Vec and the 3-tuple in the HashMap::V hurt readability.
Further, the 3-tuple in the HashMap::V has two components with the same type, which increases the risk of accidentally assigning to the wrong tuple component.

Would have preferred a cleanup that switches to a named function-scoped struct as part of this PR.

Reply:
Agree that this deserves a clean up, but let's do it separately.  We may be able to simplify it further beyond just using nicer types